### DISCUSSION TITLE
Unable to pause/resume file upload.

### DISCUSSION THREAD

**Discussion by puneesh** [aws/aws-sdk-java-v2#3781] [Q&A]
I am following samples mentioned [here ](https://sdk.amazonaws.com/java/api/latest/software/amazon/awssdk/transfer/s3/S3TransferManager.html)to pause and resume file upload to S3 bucket.

Issue I am facing is that the moment `upload.pause()` is called `java.util.concurrent.CancellationException: null`  is thrown and `Transfer failed.` is printed in log.

So, if I catch exception - I am able to persist upload information using `resumableFileUpload.serializeToFile(path)` but resume is not happening when this line executes `FileUpload resumedUpload = transferManager.resumeUploadFile(persistedResumableFileUpload)`. What am I missing here? please advise.

I am using 2.20.0 version of `s3-transfer-manager, aws-crt, aws-crt-client `under software.amazon.awssdk

**Comment by debora-ito** [comment]
When resume is called, is there any error in the logs?
Can you provide a self-contained code sample I can use to reproduce?

**Reply by puneesh** [reply]
hello @debora-ito, exception occurs when `upload.pause()` is called. I've attached code sample, please see if it can help replicate issue for you.
[PauseAndResumePOC.txt](https://github.com/aws/aws-sdk-java-v2/files/10803789/PauseAndResumePOC.txt)

**Reply by robertbio** [reply]
I have a similar use case where `CancellationException: null ` is thrown in the transfer listener when I try to pause after a network interruption during an upload/download. The uploaded file has to be big enough; once it starts uploading, just remove the internet connection and wait for the code get to `fileUpload.pause()`. To that matter, changing from wifi to cable or stopping a VPN connections also freezes the progress. One issue is the exception on pause; another is the resilience to switching connections. This is a desktop client app.
```java

	void uploadExample(S3TransferManager transferManager, String bucket, String key, Path bigLocalFilePath) throws InterruptedException, ExecutionException {
		// Upload a file to S3 usint the TransferManager.
		FileUpload fileUpload = transferManager.uploadFile(u -> u.source(bigLocalFilePath)
				.putObjectRequest(p -> p.bucket(bucket)
						.key(key))
				.addTransferListener(CustomLoggingTransferListener.create(10)));

		long lastCheckTransferedBytes = 0L;
		while (true) {
			try {
				// I had some problems where fileUpload.completionFuture().join() would get stuck if there was a connection switch,
				// so now I check progress every 5 minutes or pause and resume (and additional logic can go here).
				fileUpload.completionFuture()
						.get(5, TimeUnit.SECONDS); // .get(5, TimeUnit.MINUTES); // set seconds to fail faster.
				return; // done downloading.
			} catch (TimeoutException e) {
				long newMeasurement = fileUpload.progress().snapshot().transferredBytes();
				log.debug("Timeout: New measurement: {} - Old: {}", newMeasurement, lastCheckTransferedBytes);
				if (newMeasurement <= lastCheckTransferedBytes) {
					// no progress
					ResumableFileUpload pause = fileUpload.pause();
					// resilience logic... pause and set continue.
				}
				lastCheckTransferedBytes = newMeasurement;
			}
		}
	}

```

Edit: I corrected the code. I only see the cancellation exception in the `CustomLoggingTransferListener` that get the method `transferFailed` triggered and prints the `context.exception()`, but the pause content is ok and can be used for resuming the upload.

**Reply by debora-ito** [reply]
@puneesh Thank you for the code sample, I'll try to reproduce. 
Can you share the full stacktrace with the error you are seeing?

**Reply by puneesh** [reply]
@debora-ito please see if attached screenshot is helpful.
![IMG-6957](https://user-images.githubusercontent.com/4569911/220881073-d28cd867-c6a9-479e-b43b-a0d00f54f334.jpg)

**Reply by puneesh** [reply]
@debora-ito please advise.

**Reply by puneesh** [reply]
@debora-ito please advise, if you were able to replicate this issue.
